packages/tracing/src/browser/request.ts

```diff
@@ -0,0 +1,237 @@
+import { addInstrumentationHandler, isInstanceOf, isMatchingPattern } from '@sentry/utils';
+
+import { Span } from '../span';
+
+import { getActiveTransaction } from './utils';
+
+export const DEFAULT_TRACING_ORIGINS = ['localhost', /^\//];
+
+/** Options for Request Instrumentation */
+export interface RequestInstrumentationOptions {
+  /**
+   * List of strings / regex where the integration should create Spans out of. Additionally this will be used
+   * to define which outgoing requests the `sentry-trace` header will be attached to.
+   *
+   * Default: ['localhost', /^\//] {@see DEFAULT_TRACING_ORIGINS}
+   */
+  tracingOrigins: Array<string | RegExp>;
+
+  /**
+   * Flag to disable patching all together for fetch requests.
+   *
+   * Default: true
+   */
+  traceFetch: boolean;
+
+  /**
+   * Flag to disable patching all together for xhr requests.
+   *
+   * Default: true
+   */
+  traceXHR: boolean;
+
+  /**
+   * This function will be called before creating a span for a request with the given url.
+   * Return false if you don't want a span for the given url.
+   *
+   * By default it uses the `tracingOrigins` options as a url match.
+   */
+  shouldCreateSpanForRequest?(url: string): boolean;
+}
+
+/** Data returned from fetch callback */
+interface FetchData {
+  args: any[];
+  fetchData: {
+    method: string;
+    url: string;
+    // span_id
+    __span?: string;
+  };
+  startTimestamp: number;
+  endTimestamp?: number;
+}
+
+/** Data returned from XHR request */
+interface XHRData {
+  xhr?: {
+    __sentry_xhr__?: {
+      method: string;
+      url: string;
+      status_code: number;
+      data: Record<string, any>;
+    };
+    __sentry_xhr_span_id__?: string;
+    __sentry_own_request__: boolean;
+    setRequestHeader?: Function;
+  };
+  startTimestamp: number;
+  endTimestamp?: number;
+}
+
+export const defaultRequestInstrumentionOptions: RequestInstrumentationOptions = {
+  traceFetch: true,
+  traceXHR: true,
+  tracingOrigins: DEFAULT_TRACING_ORIGINS,
+};
+
+/** Registers span creators for xhr and fetch requests  */
+export function registerRequestInstrumentation(_options?: Partial<RequestInstrumentationOptions>): void {
+  const { traceFetch, traceXHR, tracingOrigins, shouldCreateSpanForRequest } = {
+    ...defaultRequestInstrumentionOptions,
+    ..._options,
+  };
+
+  // We should cache url -> decision so that we don't have to compute
+  // regexp everytime we create a request.
+  const urlMap: Record<string, boolean> = {};
+
+  const shouldCreateSpan = shouldCreateSpanForRequest
+    ? shouldCreateSpanForRequest
+    : (url: string) => {
+        if (urlMap[url]) {
+          return urlMap[url];
+        }
+        const origins = tracingOrigins;
+        const decision =
+          origins.some((origin: string | RegExp) => isMatchingPattern(url, origin)) &&
+          !isMatchingPattern(url, 'sentry_key');
+        urlMap[url] = decision;
+        return urlMap[url];
+      };
```

Comment:
```suggestion
  const shouldCreateSpan = shouldCreateSpanForRequest || (url: string) => {
        if (urlMap[url]) {
          return urlMap[url];
        }
        const origins = tracingOrigins;
        const decision =
          origins.some((origin: string | RegExp) => isMatchingPattern(url, origin)) &&
          !isMatchingPattern(url, 'sentry_key');
        urlMap[url] = decision;
        return urlMap[url];
      };
```

Reply:
Hmm typescript doesn't like the arrow function + or operator, so I did something like this:

```ts
const defaultShouldCreateSpan = (url: string): boolean => {
  if (urlMap[url]) {
    return urlMap[url];
  }
  const origins = tracingOrigins;
  urlMap[url] =
    origins.some((origin: string | RegExp) => isMatchingPattern(url, origin)) &&
    !isMatchingPattern(url, 'sentry_key');
  return urlMap[url];
};

const shouldCreateSpan = shouldCreateSpanForRequest || defaultShouldCreateSpan;
```